Better tests for transaction conflict handling

LagginTimes opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

Fixes bitcoindevkit/bdk#1063.

This PR introduces a new `TxTemplate` struct to test different transaction conflict scenarios in `TxGraph`.
The following transaction conflict scenarios are tested:
- 2 unconfirmed txs with different last_seens conflict. The most recent tx should be the only tx that appears in the list methods.
- 3 unconfirmed txs with different last_seens conflict. The most recent tx should be the only tx that appears in the list methods.
- An unconfirmed tx U conflicts with a tx anchored in orphaned block O. O has higher last_seen. O should be the only tx that appears in the list methods.
- An unconfirmed tx U conflicts with a tx anchored in orphaned block O. U has higher last_seen. U should be the only tx that appears in the list methods.
- Multiple unconfirmed txs conflict with a confirmed tx. None of the unconfirmed txs should appear in the list methods.
- B and B' conflict. C spends B. B' is anchored in best chain. B and C should not appear in the list methods.
- B and B' conflict. C spends B. B is anchored in best chain. B' should not appear in the list methods.
- B and B' conflict. C spends both B and B'. C is impossible.
- B and B' conflict. C spends both B and B'. C is impossible. B' is confirmed.
- B and B' conflict. C spends both B and B'. C is impossible. D spends C.

These tests revealed that `TxGraph::walk_conflicts` was not checking ancestors of the root tx for conflicts. `TxGraph::walk_conflicts` has been refactored to check for conflicting ancestor transactions by using a new `TxAncestors` iterator in `TxGraph`.

### Changelog notice

- Introduced `tx_template` module
- Introduced `TxGraph::TxAncestors` iterator
- Refactored `TxGraph::walk_conflicts` to use `TxGraph::TxAncestors`
- Added `walk_ancestors` to `TxGraph`

### Checklists
All Submissions:

- [x] I've signed all my commits
- [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
- [x] I ran cargo fmt and cargo clippy before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,131 @@
+pub struct TxTemplate<'a, K, A> {
+    /// Uniquely identifies the transaction, before it can have a txid.
+    pub tx_name: &'a str,
+    pub spends: &'a [TxInTemplate<'a>],
```
Lets rename this to `inputs`.

LagginTimes replied:
Done.

evanlinjin commented on crates/chain/tests/test_indexed_tx_graph.rs:
```diff
@@ -1,19 +1,597 @@
+}
+
+#[test]
+fn test_scenarios() {
```
I think Raj had comments explaining what was being tested. We still need them. Also the test name should be changed back to what it was (or have something better).

evanlinjin commented on crates/chain/tests/test_indexed_tx_graph.rs:
```diff
@@ -1,19 +1,597 @@
+        );
+        graph = temp_graph;
+        tx_ids.extend(temp_tx_ids);
+    }
```
Looks like we aren't using `Scenario` for the purpose that they are designed for.

If there is no way to use it for these tests here, we can just not use it.

LagginTimes replied:
This was my next step, and is what I am currently working on.

evanlinjin reviewed (commented):
Thanks for working on this, glad to see the progress. Just some comments!

evanlinjin commented on crates/chain/tests/test_indexed_tx_graph.rs:
```diff
@@ -1,19 +1,331 @@
+    }];
+
+    // A helper lambda to extract and filter data from the graph.
+    let fetch =
```
We don't need a separate closure here since we are only calling it in one place, but I guess this is a placeholder test so it'll be removed anyway?

evanlinjin commented on crates/chain/tests/test_indexed_tx_graph.rs:
```diff
@@ -1,19 +1,379 @@
+            let ((_, script), _) = graph.index.reveal_next_spk(&TestKeychain::Internal);
+            untrusted_spks.push(script.to_owned());
+        }
+    }
```
Try make use of `Iterator` and `Iterator::collect` here.

evanlinjin commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,139 @@
+#[allow(dead_code)]
+pub fn init_graph<'a>(
+    graph: IndexedTxGraph<ConfirmationHeightAnchor, KeychainTxOutIndex<TestKeychain>>,
+    descriptors: HashMap<TestKeychain, Descriptor<DescriptorPublicKey>>,
```
I just realized we probably don't need to have multiple descriptors here. Since we are testing conflicts. At most, we just need to know whether the spk is tracked or not tracked (so a single descriptor is fine).

`KeychainTxOutIndex<()>` will do I think.

LLFourn replied:
Why do we need descriptors at all? You can just insert script pubkeys into `SpkTxOutIndex`. I am a little unsure as to why we need an indexed graph either. Doesn't transaction conflict handling exclusively in the domain of methods on `TxGraph` which are passed a `ChainOracle`.

evanlinjin replied:
~@LLFourn Good point.~

Edit:

Wait, we need to descriptor, because of how `TxOutTemplate` works.

https://github.com/bitcoindevkit/bdk/blob/a86f76ae0cfd3dd333699c8c0bf70a72723cc3d0/crates/chain/tests/common/tx_template.rs#L55-L58

But we should shove the spks into `SpkTxoutIndex`. @LagginTimes also, we don't need to `K` generic anymore.

evanlinjin commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,138 @@
+
+#[allow(dead_code)]
+pub fn init_graph<'a>(
+    graph: IndexedTxGraph<ConfirmationHeightAnchor, KeychainTxOutIndex<()>>,
```
Nit: I think it will be clearer if we just had a `descriptor` input, and not construct an `IndexedTxGraph` from an `IndexedTxGraph`.

evanlinjin commented on crates/chain/tests/test_indexed_tx_graph.rs:
```diff
@@ -1,19 +1,442 @@
+            exp_chain_txs: HashSet::from([
+                ("tx1", 0),
+                ("tx_bestchain_conflict_2", 0),
+                ("tx_bestchain_spend_1", 0),
```
Let's keep the tests accurate (even though there may be a bug in `TxGraph`).

evanlinjin commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,145 @@
+    PrevTx(&'a str, usize),
+}
+
+pub struct TxOutTemplate<K> {
```
Please get rid of this generic!

LagginTimes replied:
`<K>`.

evanlinjin commented:
Please rebase on `master` to fix the CI.

evanlinjin commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,145 @@
+                let _ = graph.insert_seen_at(tx.txid(), seen_at);
+            }
+        };
+    }
```
Let's not filter out transactions. If the transaction is in the scenario, shove it in the graph.

evanlinjin commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -14,7 +14,7 @@ use crate::{
 ///
 /// This structure ensures that [`TxGraph`] and [`Indexer`] are updated atomically.
-#[derive(Debug)]
+#[derive(Clone, Debug)]
```
Why is this change necessary? Ideally we want to avoid cloning `IndexedTxGraph` at it may become problematic.

LagginTimes replied:
This was probably left over from old code/debugging use and has been removed.

evanlinjin commented:
It's looking good so far. We've discussed getting rid of the recursion. After that is done, let's clean up the commit messages:

```
test(chain): add `TxTemplate` tests...  # add a proper body & comment out failing tests
fix(chain): fix ...                     # enable the now-passing tests
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -655,25 +657,47 @@ impl<A: Anchor> TxGraph<A> {
-                if chain.is_block_in_chain(block, chain_tip)? == Some(true) {
-                    // conflicting tx is in best chain, so the current tx cannot be in best chain!
+        while !check_tx_conflicts.is_empty() {
+            let (tx, last_seen, ancestor_height) = check_tx_conflicts.pop().unwrap();
```
**Nit:** I think this is better as we can avoid the `.unwrap()`.

```suggestion
        loop {
            let (tx, last_seen, ancestor_height) = match check_tx_conflicts.pop() {
                Some(item) => item,
                None => break,
            };
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -655,25 +657,47 @@ impl<A: Anchor> TxGraph<A> {
+                                    check_tx_conflicts.push((tx, last_seen, ancestor_height + 1));
+                                }
+                                TxNodeInternal::Partial(_) => {}
+                            };
```
**Nit:**

```suggestion
                            if let TxNodeInternal::Whole(tx) = tx_node {
                                check_tx_conflicts.push((tx, last_seen, ancestor_height + 1));
                            }
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -641,6 +641,8 @@ impl<A: Anchor> TxGraph<A> {
         chain_tip: BlockId,
         txid: Txid,
     ) -> Result<Option<ChainPosition<&A>>, C::Error> {
+        let mut check_tx_conflicts: Vec<(&Transaction, &u64, u32)> = Vec::new();
```
Add a comment as to why this is needed. Something like this:

```
// This is a stack of previous spends of `txid`. Anything that conflicts with `txid`'s ancestors also conflict with `txid`.
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -655,25 +657,47 @@ impl<A: Anchor> TxGraph<A> {
-            if conflicting_tx.last_seen_unconfirmed > *last_seen {
-                return Ok(None);
+
+            // Add this tx's direct ancestor(s) to the queue to check for conflicts, with a hard
```
```suggestion
            // Add this tx's direct ancestor(s) to the stack to check for conflicts, with a hard
```

evanlinjin reviewed (changes requested):
Thank for the excellent work. A few notes.

1. Can you please add a proper PR description? We need a short description of what we are testing, and the bug that was found and how it was fixed.

2. We've identified a shortcoming of `TxGraph::walk_conflicts`. It does not find all conflicts since it does not check conflicts of ancestors of the root tx. Please include this in the docs of `walk_conflicts`. Also make a separate ticket to fix `walk_conflicts` for good.

evanlinjin commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,141 @@
+                .iter()
+                .map(|input| match input {
+                    TxInTemplate::Bogus => TxIn {
+                        // #TODO have actual random data
```
Let's complete this!

A simple solution is to use the current unix timestamp as the seed (which you can hash/modulo).

Also add some bogus inputs to test transactions.

evanlinjin reviewed (changes requested):
Last thing.

evanlinjin reviewed (approved):
ACK 6bd514518d90c221d33e0cbad4e5db5b9476510d

However, we should consider changing the behavior of `TxGraph::walk_conflicts` to walk ancestors to a given depth and check conflicts on those ancestors. This can be a separate issue.

danielabrozzoni commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,141 @@
+    fn clone(&self) -> Self {
+        *self
+    }
+}
```
Clone and Copy can be derived I think :)

evanlinjin replied:
But `A` will need to implement Copy and Clone, which most anchors do anyway? I don't mind either way.

danielabrozzoni replied:
That's the case already I think?

danielabrozzoni commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,141 @@
+                    },
+                    TxInTemplate::PrevTx(prev_name, prev_vout) => {
+                        let prev_txid = tx_ids.get(prev_name).expect(
+                            "txin template must spend from tx of template that comes before",
```
nit: It's good that there's an expect here, but maybe it should be written in the method docs as well

danielabrozzoni commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -153,89 +153,89 @@ fn test_tx_conflict_handling() {
         ..Default::default()
     };
 
-    // let tx_bestchain_conflict_1: TxTemplate<'_, ConfirmationHeightAnchor> = TxTemplate {
```
(in 1b2d552ba5333fb011125c26fd1d95190a7a425f and 6e75d051f5fd9c9e61d189893d18234ba01c86fc): sorry for being picky about commit hygiene, but I don't think we should have a commit adding a commented test and another commit uncommenting it. You can either add the test in the first commit without comments, or add it in the second commit along with the fix

evanlinjin replied:
@danielabrozzoni is there a document/article you refer to for best practice for commit hygiene? Will be helpful for future reference.

danielabrozzoni replied:
I really like https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches

But in general, for big PRs it makes sense to review each commit separately, so you should make sure that commits are not modifying what's done in previous commits in the same PR

danielabrozzoni commented on crates/chain/tests/test_indexed_tx_graph.rs:
```diff
@@ -1,473 +0,0 @@
-#[macro_use]
```
In 9aa22dd62554c0138928b14f5f4332ed040de369: why is this being removed? Can we have the reason in the commit message? Are you moving these tests somewhere else or just completely rewriting it?

LagginTimes replied:
It was decided that the file name should be renamed to `test_tx_graph_conflicts`. The old one, `test_indexed_tx_graph` was therefore removed.

LLFourn replied:
`test_index_tx_graph.rs` should have tests that specifically test things the indexed tx graph does. i.e. after you insert a tx into it, the indexer has been called and asked to index that tx. e.g. `insert_relevant_txs` should be in here since it has nothing to do with conflicts as far as I can tell.

danielabrozzoni commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -1,9 +1,10 @@
+use rand::distributions::{Alphanumeric, DistString};
```
nit: I would squash 760eaa246dbb1fc4329f7d32b00f86a0d0b9231e into 1b2d552ba5333fb011125c26fd1d95190a7a425f, but if you prefer to keep them separate it's fine as well

danielabrozzoni commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,141 @@
+
+pub struct TxOutTemplate {
+    pub value: u64,
+    pub spk_index: Option<u32>, // some = get spk from SpkTxOutIndex, none = random spk
```
I'm curious, why are you inserting the index instead of the whole spk? I'd say not to worry too much about cloning since it's just tests, clarity should be preferred.

evanlinjin replied:
Thanks for the comment! The thinking is that it is meant to be a "template" so it should be easy to read and write out. An index is perfect for this, instead of having the full spk.

danielabrozzoni commented on crates/chain/src/tx_graph.rs:
```diff
@@ -339,6 +339,9 @@ impl<A> TxGraph<A> {
     /// Creates an iterator that both filters and maps conflicting transactions (this includes
     /// descendants of directly-conflicting transactions, which are also considered conflicts).
     ///
+    /// Note that this does not find all conflicts, because we are not checking ancestors of the
```
nit in the commit message of 6e75d051f5fd9c9e61d189893d18234ba01c86fc: you have a duplicated line "fix(chain): add check for ancestor tx conflicts in `TxGraph`" at the end

danielabrozzoni commented on crates/chain/src/tx_graph.rs:
```diff
@@ -319,6 +319,12 @@ impl<A> TxGraph<A> {
     }
 
+    /// Creates an iterator of ancestor transactions from the starting `Transaction`, up to a
+    /// specified depth limit.
```
nit: we should document that this doesn't include the tx itself. I know it seems obvious but it might not be!

danielabrozzoni commented on crates/chain/src/tx_graph.rs:
```diff
@@ -357,7 +368,8 @@ impl<A> TxGraph<A> {
     /// Note that this only returns directly conflicting txids and does not include descendants of
-    /// those txids (which are technically also conflicting).
+    /// those txids or conflicting txids of the given transaction's ancestors (which are technically
+    /// also conflicting).
```
nit: I would rewrite it as:
```
/// Note that this only returns directly conflicting txids and does not include:
/// - descendants of conflicting transactions (which are technically also conflicting)
/// - transactions conflicting with the given transaction's ancestors
```

evanlinjin replied:
"may not include"

danielabrozzoni commented on crates/chain/src/tx_graph.rs:
```diff
@@ -674,7 +686,7 @@ impl<A: Anchor> TxGraph<A> {
         // ignore it by checking conflicts!
         let tx = match tx_node {
-            TxNodeInternal::Whole(tx) => tx,
+            TxNodeInternal::Whole(tx) => tx, // check_tx_conflicts.push((tx, last_seen, 0)),
```
Is this supposed to be here?

danielabrozzoni commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1110,6 +1122,118 @@ impl<A> ForEachTxOut for TxGraph<A> {
+pub struct TxAncestors<'g, A> {
+    graph: &'g TxGraph<A>,
+    visited: HashSet<Txid>,
+    stack: Vec<(usize, Transaction)>,
```
nit: document what this usize is (depth?)

evanlinjin replied:
```suggestion
    stack: Vec<(usize, &'g Transaction)>,
```

The iterator should use references of transactions instead of cloning them. You'll need to play around with lifetimes to make this happen.

danielabrozzoni commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1110,6 +1122,118 @@ impl<A> ForEachTxOut for TxGraph<A> {
+                .iter()
+                .map(|txin| txin.previous_output.txid)
+                .for_each(|prev_txid| {
+                    if let Some((TxNodeInternal::Whole(tx), _, _)) = self.graph.txs.get(&prev_txid)
```
Do I understand correctly that we don't traverse partial txs? If that's the case, we should document it

danielabrozzoni commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -496,6 +496,209 @@ fn test_calculate_fee_on_coinbase() {
 
+// `test_walk_ancestors` uses the following transaction structure:
+//
+//     a0
+//    /  \
+//   b0   b1   b2
+//  /  \   \  /
+// c0  c1   c2  c3
+//    /      \  /
+//   d0       d1
+//             \
+//              e0
```
I know Evan praised you already, but I feel like repeating it - thank you for this sketch of the txgraph, it is truly helpful!

danielabrozzoni commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -496,6 +496,209 @@ fn test_calculate_fee_on_coinbase() {
+
+    let mut ancestors = vec![
+        graph
+            .walk_ancestors(tx_e0.clone(), 25)
```
TxAncestors is already an iterator, you don't need to collect and iter again:
```
        graph
            .walk_ancestors(tx_e0.clone(), 25)
            .map(|(depth, tx)| (depth, tx.txid()))
            .collect::<Vec<_>>(),
```

danielabrozzoni commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -496,6 +496,209 @@ fn test_calculate_fee_on_coinbase() {
+    ];
+
+    while let Some(mut txids) = ancestors.pop() {
+        let mut expected_txids = expected_ancestors.pop().unwrap();
```
Commenting just in case you ever need it, you can just zip the two vectors in this case. But popping works just fine!

evanlinjin replied:
@LagginTimes I would go ahead and use [`zip`](https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.zip) here.

Another trick would be to collect into a `BTreeSet` so you don't need to `.sort()`.

danielabrozzoni commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1282 @@
+            },
+        },
+        Scenario {
+            name: "unconfirmed tx conflicts with tx in orphaned block, orphaned higher unseen",
```
Here and below, I think you meant `last_seen` instead of unseen?

danielabrozzoni commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1282 @@
+        anchors: &[ConfirmationHeightAnchor {
+            anchor_block: BlockId {
+                height: 5,
+                hash: h!("Orphaned Block"),
```
Btw, I think the orphaned block should have height = 4, as we don't have a block with height 5 in the best chain. But I don't think it really matters for the test scenario

danielabrozzoni commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1282 @@
+            },
+        },
+        Scenario {
+            name: "B and B' conflict, C spends B, B' is anchored in best chain",
```
You should specify that B and B' both spend from A, which is in the best chain

danielabrozzoni commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1282 @@
+/// This test also checks that [`TxGraph::filter_chain_txouts`], [`TxGraph::filter_chain_unspents`],
+/// and [`TxGraph::balance`] return correct data.
+#[test]
+fn test_tx_conflict_handling() {
```
nit: this test is quite difficult to follow as there are a lot of test scenarios, and I found myself going back and forth between the tx declaration and the test cases; it would be helpful to split it into smaller tests.

evanlinjin replied:
I agree with this. I think we should declare the tx templates in the actual scenarios.

To make it more readable, we can also do the following:
* Use `BlockId` as the anchor (maybe add a `block_id!(1, "B")` macro so each anchor declaration is on one line).
* Add `TxOutTemplate::new(value, spk_index)` method, or a macro. This way each txout declaration is on one line.

danielabrozzoni reviewed (commented):
Thanks a lot for your work! This is just a partial review, I reviewed everything apart from the tests in `crates/chain/tests/test_tx_graph_conflicts.rs`, I only reviewed `test_tx_conflict_handling` in that file, I'm going to finish the review in the next few days. 

#1063 includes a "Further work" section that I don't think is addressed here, am I correct? If that's the case, we should open a new issue before merging this PR

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -319,6 +319,12 @@ impl<A> TxGraph<A> {
 
+    /// Creates an iterator of ancestor transactions from the starting `Transaction`, up to a
+    /// specified depth limit.
+    pub fn walk_ancestors(&self, tx: Transaction, depth_limit: usize) -> TxAncestors<A> {
```
```suggestion
    pub fn walk_ancestors<'g>(&'g self, tx: &'g Transaction, depth_limit: usize) -> TxAncestors<'g, A> {
```

also make `TxAncestors` iterate references to `Transaction`s.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -337,18 +343,23 @@ impl<A> TxGraph<A> {
+        let mut txids = HashSet::new();
+        for (_, tx) in TxAncestors::new_include_root(self, tx.clone(), depth_limit) {
+            txids.extend(self.direct_conflicts_of_tx(&tx).map(|(_, txid)| txid));
+        }
```
Nit: this can be written in a more "functional" manner.

evanlinjin commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1282 @@
+    /// Name of the test scenario
+    name: &'a str,
+    /// Transaction templates
+    tx_templates: &'a [TxTemplate<'a, ConfirmationHeightAnchor>],
```
Since #1069 is merged now, we should use `BlockId` as the anchor to simplify these tests.

evanlinjin commented:
@danielabrozzoni bitcoindevkit/bdk#1102 is the ticket for the "further work" section.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -683,7 +695,7 @@ impl<A: Anchor> TxGraph<A> {
         // If a conflicting tx is in the best chain, or has `last_seen` higher than this tx, then
         // this tx cannot exist in the best chain
-        for conflicting_tx in self.walk_conflicts(tx, |_, txid| self.get_tx_node(txid)) {
+        for conflicting_tx in self.walk_conflicts(tx, 25, |_, txid| self.get_tx_node(txid)) {
```
I think it would make sense to make the `depth_limit` input configurable.

However, because `try_get_chain_position` is called in so many places, would it make sense to make all these methods take in an `ancestor_depth_limit` argument?

The other option is to have the `ancestor_depth_limit` global for the entire `TxGraph`, but would there be a situation where we want different depth limits for the same graph?

Is there even a way to avoid the depth limit entirely?

LLFourn replied:
I think as long as we accept this `25` is a temporary solution it's fine to leave it fixed. Let's put it in a constant.

LLFourn replied:
Rather than just going to `25` it would be better to continue getting ancestor conflicts until you reach the first confirmed ancestor because then it no longer matters what its conflicts are. I'm not sure if this be done easily.

I wrote what I think the long term plan to approach the general problem should be in: https://github.com/bitcoindevkit/bdk/issues/1099#issuecomment-1704559684 and remove this limit anyway.

LagginTimes commented:
I've addressed all requested changes except for the ones related to commit hygiene, which will be addressed at a later date.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -337,27 +347,35 @@ impl<A> TxGraph<A> {
     pub fn walk_conflicts<'g, F, O>(
         &'g self,
         tx: &'g Transaction,
+        depth_limit: usize,
```
I think that we don't want to add this parameter here. I see why this approach needs this limit but I don't think it belongs in the public API. `walk_conflicts` should let you walk all the transactions that would have to not exist for this transaction to be canonical assuming the parents of this tx are canonical. This makes sense in the context of broadcasting a transaction: you want to be able to find out which txs this will remove when you broadcast it and it is accepted. In this case you already know that all the parents are canonical (otherwise your broadcast will fail anyway).

I think that you can still do this full conflict list (including transitive ancestor conflicts) with internal logic which just calls walk conflicts for each ancestor up to a certain depth.

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1348 @@
+        2
+    );
+
+    // We have two mempool utxos both at same chain position.
```
why do we assert the same thing twice.

LagginTimes replied:
This was code that was incorporated from #964 and has not been refactored with new `TxTemplate` yet. My PR description should have mentioned the fact that almost all of the code taken from #964 has thus far been left as-is.

evanlinjin replied:
@LagginTimes let's extract what is good from #964 into "scenarios".

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1348 @@
+    // Initiate IndexedTxGraph
+
+    let (desc_1, _) = Descriptor::parse_descriptor(&Secp256k1::signing_only(), "tr(tprv8ZgxMBicQKsPd3krDUsBAmtnRsK3rb8u5yi1zhQgMhF1tR8MW7xfE4rnrbbsrbPR52e7rKapu6ztw1jXveJSCGHEriUGZV7mCe88duLp5pj/86'/1'/0'/0/*)").unwrap();
+    let (desc_2, _) = Descriptor::parse_descriptor(&Secp256k1::signing_only(), "tr(tprv8ZgxMBicQKsPd3krDUsBAmtnRsK3rb8u5yi1zhQgMhF1tR8MW7xfE4rnrbbsrbPR52e7rKapu6ztw1jXveJSCGHEriUGZV7mCe88duLp5pj/86'/1'/0'/1/*)").unwrap();
```
This test looks like it's testing aspects of `IndexedTxGraph` so it should be in `test_indexed_tx_graph.rs`.

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1348 @@
+/// are associated with owned spks in the index). Ensure insertion and indexing is topological-
+/// agnostic.
+#[test]
+fn insert_relevant_txs() {
```
This doesn't seem to be about testing conflicts but it's in `test_tx_graph_conflicts.rs`.

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1348 @@
+}
+
+#[allow(unused)]
+pub fn single_descriptor_setup() -> (
```
we shouldn't need an indexed tx graph to test conflicts.

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1348 @@
+    name: &'a str,
+    /// Transaction templates
+    tx_templates: &'a [TxTemplate<'a, BlockId>],
+    /// Names of txs that must exist in the output of `list_chain_txs`
```
the output of `list_chain_txs` is never tested anywhere. Can we test it like this comment implies we should?

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,1348 @@
+
+/// Check conflicts between two in mempool transactions. Tx with older `seen_at` is filtered out.
+#[test]
+fn test_unconfirmed_conflicts() {
```
Is there a reason this test and the others that use `setup_conflicts` cannot be incorporated into the "scenarios" structure above?

LagginTimes replied:
This is also mostly unmodified code from #964. In discussion with @evanlinjin, one of the next steps was to refactor the #964 tests with the `TxTemplate` scenario structure.

LLFourn reviewed (changes requested):
Hey @LagginTimes. Thanks for looking at how to make the conflict handling better. I your approach is the simplest way to get the tests passing to fix the intended bug. Great work. In practice the implementation might be inefficient with a deep mempool and lots of conflicting txs because of the repeated ancestor traversal. In https://github.com/bitcoindevkit/bdk/issues/1099#issuecomment-1704559684 I put forward a plan to solve this. Having said that, getting tests passing is more important for now and we can iterate on that later.

I've left comments explaining how this PR can be improved but in summary:

1. Leave the public API the same (except for adding the lovely new ancestors traversal feature)
2. In the tx graph conflict tests, don't use an indexed tx graph.
3. restore `test_indexed_tx_graph.rs` and put things in there that test its API (e.g. `insert_relevant_txs`.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -319,14 +319,19 @@ impl<A> TxGraph<A> {
-    /// Creates an iterator of ancestor transactions from the starting `Transaction` up to a
-    /// specified depth limit, not including the original `Transaction`.
-    pub fn walk_ancestors<'g>(
+    /// Creates an iterator of unconfirmed ancestor transactions from the starting `Transaction`,
```
`walk_ancestors` should traverse all ancestors. It sounds like a general method traversing ancestors in a graph. You could have another method for `walk_unconfirmed_ancestors`.

LagginTimes replied:
The next step, after discussion with @evanlinjin, was to refactor `TxAncestors` again to take in a closure for versatility (similar to `TxDescendants`), which will allow `walk_ancestors` to traverse all ancestors. The work for this is nearly complete.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -205,6 +205,26 @@ impl<A> TxGraph<A> {
     }
 
+    /// Get a transaction by txid. This only returns `Some` for full, unconfirmed transactions.
+    pub fn get_unconfirmed_tx<'g, C>(
```
Try not to introduce new API if it's not needed. This method is only used internally, and as a public API, `get_chain_position` can be used instead.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -205,6 +205,26 @@ impl<A> TxGraph<A> {
+        A: Anchor,
+        C: ChainOracle,
+    {
+        let tx_node = self.get_tx_node(txid).expect("tx must exist");
```
Given that the intention of this method is to be a public API, we should not panic if tx does not exist, since the caller provides the `txid`.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -699,18 +725,24 @@ impl<A: Anchor> TxGraph<A> {
+        // Populate a list of unconfirmed ancestor txs to check for conflicts
+        for tx in TxAncestors::new_include_root(self, tx, |_, tx: &Transaction| {
+            self.get_unconfirmed_tx(tx.txid(), chain, chain_tip)
+                .expect("error is infallible")
```
We cannot guarantee that the `ChainOracle` implementation has an [`Infallible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html) error type.

A solution is to have the closure return `Option<Result<&tx, C::Error>>`, and we collect the ancestor txs into `Result<Vec<_>, C::Error>`. You'll need to play around with the lifetimes.

evanlinjin reviewed (changes requested):
Good work so far.

Please remove `TxGraph::get_unconfirmed_tx` and make sure we aren't panicing in unexpected ways.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -681,11 +705,29 @@ impl<A: Anchor> TxGraph<A> {
+        for conflicting_tx in ancestor_txs
+            .iter()
+            .flat_map(|&tx| self.walk_conflicts(tx, |_, txid| self.get_tx_node(txid)))
+        {
```
Really small nit: for clarity, I would declare the iterator as a variable. Also, we can use `into_iter` to consume `ancestor_txs` since we aren't using it anymore.
```suggestion
        let conflicting_txs = ancestor_txs
            .into_iter()
            .flat_map(|ancestor| self.walk_conflicts(ancestor, |_, txid| self.get_tx_node(txid)));

        for conflicting_tx in conflicting_txs {
```

evanlinjin commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,583 @@
+            balance, scenario.exp_balance,
+            "\n[{}] 'balance' failed",
+            scenario.name
+        );
```
Nit: I think this will make it easier to read.

```suggestion
        let txs = tx_graph
            .list_chain_txs(&local_chain, chain_tip)
            .map(|tx| tx.tx_node.txid)
            .collect::<BTreeSet<_>>();
        let exp_txs = scenario
            .exp_chain_txs
            .iter()
            .map(|txid| *exp_tx_ids.get(txid).expect("txid must exist"))
            .collect::<BTreeSet<_>>();
        assert_eq!(
            txs, exp_txs,
            "\n[{}] 'list_chain_txs' failed",
            scenario.name
        );

        let txouts = tx_graph
            .filter_chain_txouts(
                &local_chain,
                chain_tip,
                spk_index.outpoints().iter().cloned(),
            )
            .map(|(_, full_txout)| full_txout.outpoint)
            .collect::<BTreeSet<_>>();
        let exp_txouts = scenario
            .exp_chain_txouts
            .iter()
            .map(|(txid, vout)| OutPoint {
                txid: *exp_tx_ids.get(txid).expect("txid must exist"),
                vout: *vout,
            })
            .collect::<BTreeSet<_>>();
        assert_eq!(
            txouts, exp_txouts,
            "\n[{}] 'filter_chain_txouts' failed",
            scenario.name
        );

        let utxos = tx_graph
            .filter_chain_unspents(
                &local_chain,
                chain_tip,
                spk_index.outpoints().iter().cloned(),
            )
            .map(|(_, full_txout)| full_txout.outpoint)
            .collect::<BTreeSet<_>>();
        let exp_utxos = scenario
            .exp_unspents
            .iter()
            .map(|(txid, vout)| OutPoint {
                txid: *exp_tx_ids.get(txid).expect("txid must exist"),
                vout: *vout,
            })
            .collect::<BTreeSet<_>>();
        assert_eq!(
            utxos, exp_utxos,
            "\n[{}] 'filter_chain_unspents' failed",
            scenario.name
        );

        let balance = tx_graph.balance(
            &local_chain,
            chain_tip,
            spk_index.outpoints().iter().cloned(),
            |_, spk: &Script| spk_index.index_of_spk(spk).is_some(),
        );
        assert_eq!(
            balance, scenario.exp_balance,
            "\n[{}] 'balance' failed",
            scenario.name
        );
```

evanlinjin replied:
I just had an idea. I think it will be better if this is a method of `Scenario`. Then we can split these scenarios into separate test methods. This will make it easier to spot which test fails. If we do this, we can remove the `Scenario::name` field and just use the test function's name.

Let's do this is a separate PR, let's create a ticket for it.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -319,6 +319,28 @@ impl<A> TxGraph<A> {
+    /// * `ancestor_tx` is the `Transaction`'s ancestor which we are considering to walk.
+    ///
+    /// The supplied closure returns an `Option<T>`, allowing the caller to map each `Transaction`
+    /// it visitsand decide whether to visit ancestors.
```
typo

```suggestion
    /// it visits and decide whether to visit ancestors.
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -319,6 +319,28 @@ impl<A> TxGraph<A> {
     }
 
+    /// Creates an iterator of ancestor transactions from the starting `Transaction`, not including
+    /// the original `Transaction`.
```
**Nit:** I would first mention that this iterator that filters and maps. Then go on to explain where we start iterating from.

```suggestion
    /// Creates an iterator that filters and maps ancestor transactions.
    ///
    /// The iterator starts with the ancestors of the supplied `tx` (ancestor transactions of `tx`
    /// are transactions spent by `tx`). The supplied transaction is excluded from the iterator.
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -337,7 +359,8 @@ impl<A> TxGraph<A> {
     /// Creates an iterator that both filters and maps conflicting transactions (this includes
-    /// descendants of directly-conflicting transactions, which are also considered conflicts).
+    /// ancestors and descendants of directly-conflicting transactions, which are also considered
+    /// conflicts).
```
```
    /// Creates an iterator that both filters and maps directly-conflicting transactions and their
    /// descendants.
    ///
    /// Directly conflicting transactions of `tx` will have atleast one identical output spend as
    /// `tx`.
    ///
    /// **WARNING:** This will not include all conflicts as conflicts with `tx`'s ancestors and
    /// those ancestor's descendants are excluded.
    ///
    /// Refer to [`Self::walk_descendants`] for `walk_map` usage.
```

I think this is better.

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -681,11 +705,29 @@ impl<A: Anchor> TxGraph<A> {
             }
         };
 
+        // Traverse tx and its unconfirmed ancestor txs to check for conflicts
```
Code comments should explain ***why***!

```suggestion
        // Since transactions that conflict with the root tx are also conflicts, we need to traverse
        // all ancestor's conflicts. We only traverse unconfirmed ancestors since conflicts of 
        // confirmed transactions cannot be in the best chain.
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1149,6 +1191,123 @@ impl<A> ForEachTxOut for TxGraph<A> {
     }
 }
 
+/// An iterator that traverses unconfirmed ancestors that are whole `Transaction`s.
```
```suggestion
/// An iterator that traverses ancestors of a given root transaction.
///
/// The iterator excludes partial transactions.
```

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1149,6 +1191,123 @@ impl<A> ForEachTxOut for TxGraph<A> {
+            }
+        };
+        self.populate_stack(ancestor_depth + 1, tx);
+        Some(item)
```
Another nit.

```suggestion
        loop {
            // we have exhausted all paths when stack is empty
            let (ancestor_depth, tx) = self.stack.pop()?;

            // ignore paths when user filters them out
            let item = match (self.filter_map)(ancestor_depth, tx) {
                Some(item) => item,
                None => continue,
            };

            self.populate_stack(ancestor_depth + 1, tx);
            return Some(item);
        }
```

evanlinjin reviewed (changes requested):
I'm really liking how this is turning out. I have some documentation suggestions. Otherwise, it's just a bunch of nits.

evanlinjin commented:
In terms of code, I am happy with everything - good job! I'll ACK this once we fix the commit messages.

* 809a98595188c9e70dc2d4a8a3829818bea6dcb6 should be split into two. One for the `chore` of moving tests. The other one for documentation changes and other tweaks.
* 686789b9f0d004161e1d57e14582333fec76739b and 2ade5474185d6950605366546b5a3de41a47c22a should be merged (as mentioned by @danielabrozzoni).

vladimirfomene commented on crates/chain/tests/common/tx_template.rs:
```diff
@@ -0,0 +1,132 @@
+};
+use miniscript::Descriptor;
+
+/// Transaction template.
```
can we add more doc here to help anyone trying to understand the code why we decided to use this struct?

evanlinjin commented:
@LagginTimes Remember to rebase, we have conflicts!

danielabrozzoni commented:
>  I'll ACK this once we fix the commit messages.

In case you have never done that, https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History can be of help :)

evanlinjin reviewed (approved):
ACK 452627dacf4aae9e3636317223f37dae3f33397b

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -319,6 +319,30 @@ impl<A> TxGraph<A> {
 
+    /// Creates an iterator that filters and maps ancestor transactions.
+    ///
+    /// The iterator starts with the ancestors of the supplied `tx` (ancestor transactions of `tx`
```
I'm not sure this is true looking at the algorithm. If tx has ancestors `A` and `B`, it will pop off `A` from the stack and then add `A`'s ancestors to the stack. Unless I'm mistaken this means you'll visit `A`'s ancestors before visiting `B`.

The tests you've added don't test for the order they are emitted. Shouldn't they?
You could fix this in this PR or in a follow up because the problem seems to predate (it exists in `TxDescendants`).

danielabrozzoni replied:
Agreed, I fixed by using a `VecDeque` in both `TxAncestors` and `TxDescendants`. Also, the tests now check the order.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -319,6 +319,30 @@ impl<A> TxGraph<A> {
+    /// The supplied closure takes in two inputs `(depth, ancestor_tx)`:
+    ///
+    /// * `depth` is the distance between the starting `Transaction` and the `ancestor_tx`. I.e., if
+    ///    the `Transaction` is spending an output of the `ancestor_tx`; the `depth` will be 1.
```
```suggestion
    ///    the `Transaction` is spending an output of the `ancestor_tx` then `depth` will be 1.
```
^Grammatical error also exists in the `walk_descendants` function (`;` shouldn't be used to join conditional clauses; it's for independent clauses).

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -337,7 +361,8 @@ impl<A> TxGraph<A> {
 
     /// Creates an iterator that both filters and maps conflicting transactions (this includes
-    /// descendants of directly-conflicting transactions, which are also considered conflicts).
+    /// ancestors and descendants of directly-conflicting transactions, which are also considered
```
How does this walk ancestors? It only seems to walk direct conflicts and their descendants.

danielabrozzoni replied:
Agreed, I was confused as well. Fixed.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -681,11 +707,32 @@ impl<A: Anchor> TxGraph<A> {
+        // Since transactions that conflict with the root tx are also conflicts, we need to traverse
+        // all ancestor's conflicts. We only traverse unconfirmed ancestors since conflicts of
+        // confirmed transactions cannot be in the best chain.
+        let ancestor_txs =
```
```suggestion
        let unconfirmed_ancestor_txs =
```

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -356,8 +381,9 @@ impl<A> TxGraph<A> {
+    /// Note that this only returns directly conflicting txids and may not include:
+    /// - descendants of conflicting transactions (which are technically also conflicting)
+    /// - transactions conflicting with the given transaction's ancestors
     pub fn direct_conflicts_of_tx<'g>(
```
side note. This function shouldn't be suffixed `_of_tx` unless we also suffix the other method so it's `walk_conflicts_of_tx`. I think we should just remove the suffix from here.

danielabrozzoni replied:
Agreed, opened #1138 as this PR is quite big already

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -681,11 +707,32 @@ impl<A: Anchor> TxGraph<A> {
-        for conflicting_tx in self.walk_conflicts(tx, |_, txid| self.get_tx_node(txid)) {
-            for block in conflicting_tx.anchors.iter().map(A::anchor_block) {
-                if chain.is_block_in_chain(block, chain_tip)? == Some(true) {
+        for conflicting_tx in conflicting_txs {
```
I don't think the logic in this loop is right. Consider a transaction `B` with parent `A` which conflicts with `A'`. The last seen order is `B > A' > A`. Now this function will say that `B` is in the chain but `A`  is not because it conflicts with `A'` which has a higher last seen. I looked for a test for this to see if I'm wrong but it doesn't seem to be one.

danielabrozzoni replied:
Agreed, I fixed by using `direct_conflicts_of_tx` instead of `walk_conflicts`, and considering the last_seen of each ancestor instead of the last_seen of the root tx:
```diff
-        let conflicting_txs = ancestor_txs
-            .into_iter()
-            .flat_map(|ancestor| self.walk_conflicts(ancestor, |_, txid| self.get_tx_node(txid)));
-
-        // If a conflicting tx is in the best chain, or has `last_seen` higher than this tx, then
-        // this tx cannot exist in the best chain
-        for conflicting_tx in conflicting_txs {
-            for block in conflicting_tx.anchors {
-                if chain.is_block_in_chain(block.anchor_block(), chain_tip)? == Some(true) {
-                    // conflicting tx is in best chain, so the current tx cannot be in best chain!
+        for tx_node in unconfirmed_ancestor_txs {
+            // We retrieve all the transactions conflicting with this specific ancestor
+            let conflicting_txs = self
+                .direct_conflicts_of_tx(tx_node.tx)
+                .filter_map(|(_, txid)| self.get_tx_node(txid));
+
+            // If a conflicting tx is in the best chain, or has `last_seen` higher than this ancestor, then
+            // this tx cannot exist in the best chain
+            // Note that `direct_conflicts_of_tx`, which we used above to retrieve conflicts, doesn't include conflicting
+            // descendants, so comparing `last_seen` is safe
+            for conflicting_tx in conflicting_txs {
+                for block in conflicting_tx.anchors {
+                    if chain.is_block_in_chain(block.anchor_block(), chain_tip)? == Some(true) {
+                        // conflicting tx is in best chain, so the current tx cannot be in best chain!
+                        return Ok(None);
+                    }
+                }
+                if conflicting_tx.last_seen_unconfirmed > tx_node.last_seen_unconfirmed {
                     return Ok(None);
                 }
             }
-            if conflicting_tx.last_seen_unconfirmed > *last_seen {
-                return Ok(None);
-            }

```

I added the scenario "B and B' spend A and conflict, C spends B, all the transactions are unconfirmed, B' has higher last_seen than B" in the conflicting tests, which would previously fail and now passes. Lmk if it seems correct to you!

LLFourn reviewed (changes requested):
Thanks @LagginTimes. I've pointed out what I think is a bug that I think should be fixed here. The other things I've identified I leave to you to resolve them here or we can do it in follow up work.

The key things we should (eventually) resolve are:

1. The worst case time complexity of the algorithm to determine whether a single tx is in the chain is roughly linear in the number of unconfirmed transactions. If we are determining whether a bunch of txouts are in the chain and every unconfirmed tx has one this means it's quadratic. I fear by peppering the mempool with a bunch of unconfirmed txs and RBFing them a lot you could denial of service attack systems using this algorithm.
2. I think it'd be better if `TxDescendents` and `TxAncestors` should be proximity to the root txid in the graph. The current order is depth first and at what "depth" the node is visited at depends on the order of the outputs or inputs that are being traversed. The iterator can give a depth of `7` to a tx that has a direct relationship to the root tx.

danielabrozzoni committed:
test(chain): Introduce TxTemplate

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni committed:
test(chain): Add test for conflicting transactions

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni committed:
feat(chain): Introduce TxAncestors, walk_ancestors

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni committed:
doc(chain): Fix typos

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni committed:
test(chain): Add test for walk_ancestors

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni commented:
As discussed with @evanlinjin, I'm picking this up as it's high priority :)

danielabrozzoni committed:
test(chain): add block_id! utility macro

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni committed:
doc(chain): Clarify direct_conflicts_of_tx's docs

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni committed:
fix(chain): Consider conflicting ancestors in...
...try_get_chain_pos

In try_get_chain_pos, when we notice that a transaction is not included
in the best chain, we check the transactions in mempool to find
conflicting ones, and decide based on that if our transaction is still
in mempool or has been dropped.
This commit adds a check for transactions conflicting with the
unconfirmed ancestors of our tx.

Co-authored-by: Wei Chen <wzc110@gmail.com>

danielabrozzoni committed:
fix(chain): TxDescendants performs a BFS

This commit also changes test_descendants_no_repeat to check
the order of the transactions returned

danielabrozzoni reviewed (commented):
I just pushed df1ee4f495478f6131df77a4cf5dbda115923fac:
- Rewrite commit history
- Fix review comments

As a part of fixing the review comments I added a scenario in `test_tx_graph_conflicts`, I just wanted to say @LagginTimes, the `Scenario/TxTemplate` mechanism is super convenient! Good job :)

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -0,0 +1,627 @@
+                },
+            ],
+            // B and C should not appear in the list methods
+            exp_chain_txs: HashSet::from(["A", "B'"]),
```
Isn't what we want to happen here is that `B` and `C` are canonical?

`C` has the more recent last seen which should transitively mean that `B` is chosen over `B'` . In other words the effective `last_seen` of a node is the max of all its descendants.

This PR at least makes the answer consistent but it seems we are actually asserting suboptimal behavior here. Can be done in a follow up.

danielabrozzoni replied:
Just to give a bit of context, my reasoning behind this was: how would a node react if it saw B, then B', then C? And I think it would go like:
- B: fine, I'll keep this
- B': fine, I'll keep this, and drop B
- C: I'll drop this, as it's spending B, which I dropped already

Which I don't know if it's coherent with the rest of the txgraph and the last_seen logic (it probably isn't).

Anyways, I just pushed a fix, it should be fine now :)

LLFourn replied:
> Which I don't know if it's coherent with the rest of the txgraph and the last_seen logic (it probably isn't).

Indeed. TxGraph is montone so doesn't record the order of things entering into it. Our conflict resolution mechanism is `last_seen` so I think we should take whichever subgraph has the tx with the latest one. Hopefully we can make this efficient!

LLFourn reviewed (approved):
Approved. Consider fixing https://github.com/bitcoindevkit/bdk/pull/1064/files#r1344900282
ACK df1ee4f495478f6131df77a4cf5dbda115923fac

evanlinjin reviewed (approved):
ACK 6d601a7e885bfc627594b794b1a460be47799eea

Thanks @danielabrozzoni for taking this forward and @LagginTimes for the initial work.
